### DISCUSSION TITLE
How to use python client with microphone？

### DISCUSSION THREAD

**Discussion by YAETHeryk** [modelscope/FunASR#2323] [Q&A] [answered]
I deploy the service using Docker and test the sample client on my local machine. The problem now is that I test the HTML page, record with my microphone, and it works correctly. Then I test the Python client using a WAV file, and the result is also correct. 
**However**, I  continue to test the microphone mode of the Python client with command:
`python funasr_wss_client.py --host "0.0.0.0" --port 10095 --mode offline`
the client only shows:
```
Namespace(host='*******', port=10095, chunk_size=[5, 10, 5], chunk_interval=10, hotword='', audio_in=None, audio_fs=16000, send_without_sleep=True, thread_num=1, words_max_print=10000, output_dir='./results', ssl=1, use_itn=1, mode='offline')
Namespace(host='*******', port=10095, chunk_size=[5, 10, 5], chunk_interval=10, hotword='', audio_in=None, audio_fs=16000, send_without_sleep=True, thread_num=1, words_max_print=10000, output_dir='./results', ssl=1, use_itn=1, mode='offline')
connect to wss://*******:10095
```

No output, despite occupying the microphone.
I'm not sure if any steps or operations were missed.

**Comment by LauraGPT** [comment, marked as answer]
Two details in that command explain the missing output:

1. 0.0.0.0 is a server bind address, not the address a client should connect to. Use 127.0.0.1 when Docker publishes the port on the same machine, or use the Docker host's reachable IP/DNS name.
2. offline mode expects a finite utterance and produces its result after the client sends an end-of-speech message. The bundled microphone loop is continuous, so it does not create that final boundary. Use online or 2pass for a live microphone.

With the current client, install PyAudio and websockets, grant the Python process microphone permission, and omit --audio_in:

~~~bash
python -m pip install pyaudio websockets

python runtime/python/websocket/funasr_wss_client.py   --host 127.0.0.1   --port 10095   --mode 2pass   --ssl 0
~~~

Use --ssl 0 only when the deployed server uses plain ws://. If the server is configured for wss://, keep the default --ssl 1.

The microphone path captures 16 kHz, mono, signed 16-bit PCM and prints online partial text plus the corrected 2pass result at each detected sentence end. No --audio_in argument is the switch that selects this path.

If you specifically need offline recognition, first finish the recording and send it as a finite file:

~~~bash
python runtime/python/websocket/funasr_wss_client.py   --host 127.0.0.1   --port 10095   --mode offline   --audio_in recording.wav   --ssl 0
~~~

A custom live client can also obtain an offline final result by stopping capture, sending {"is_speaking": false}, and continuing to receive until is_final=true.

Current source and documentation:

- [microphone selection and PyAudio capture](https://github.com/modelscope/FunASR/blob/1d8080af99fad370645eec7de9d1bcc5ae4ab587/runtime/python/websocket/funasr_wss_client.py#L31-L52)
- [microphone PCM configuration and send loop](https://github.com/modelscope/FunASR/blob/1d8080af99fad370645eec7de9d1bcc5ae4ab587/runtime/python/websocket/funasr_wss_client.py#L122-L185)
- [official 2pass microphone command](https://github.com/modelscope/FunASR/blob/1d8080af99fad370645eec7de9d1bcc5ae4ab587/runtime/docs/SDK_tutorial_online.md#L38-L45)
